### DISCUSSION TITLE
Support IoTDB as Time Series Storage

### DISCUSSION THREAD

**Discussion by bossenti** [apache/streampipes#2857] [Dev]
I'm currently working on supporting Apache IoTDB as an additional option to store our time series data in StreamPipes.
My experience with IoTDB is limited to reading the documentation, performing some small evaluations and the implementation effort I've spent sofar.
Thus, the following is based on my current understanding of IoTDB's concept and may therefore be erroneous or incomplete.

## What is already possible

As of now, we have a very rudimentary support for Apache IoTDB as time series storage in our development branch.
Using IoTDB as time series storage is possible by setting a set of environment variables.
This effects that all time series data persisted in StreamPipes is persisted in the IoTDB.

Under the hood this works the following:

Let's assume we have created a data stream in StreamPipes with help of the [machine data simulator](https://streampipes.apache.org/docs/pe/org.apache.streampipes.connect.iiot.adapters.simulator.machine/)
and persist the data in StreamPipes into a measure called `flowrate`.
StreamPipes now has a so called `DataLakeMeasure` with the name `flowrate` and the following schema:
* `timestamp`: Long
* `sensorId`: String
* `mass_flow`: Float
* `volume_flow`: Float
* `temperature`: Float
* `density`: Float
* `sensor_fault_flags`: Boolean

Every event of our data stream is written into aligned time series of one device.
The device name in our scenario equals the name of the data lake measure (`flowrate`) and the following path is used within IoTDB: `root.streampipes.flowrate`.
This is how the data looks like:

```sql
select * from root.streampipes.flowrate limit 10 align by device
```
![image](https://github.com/apache/streampipes/assets/50115603/19931c91-1879-485e-87f6-2fb368f1c2b1)

In addition, the count mechanism to get an overview about how many events exists per data lake measure is implemented as well:
![image](https://github.com/apache/streampipes/assets/50115603/eefe4a79-0ed9-465d-9406-f490e5a31dec)

Using the following query: 
```sql
Select count(temperature) from root.streampipes.flowrate
```
![image](https://github.com/apache/streampipes/assets/50115603/032ef7ed-ff61-4b47-be61-61062e5c1212)

Well, so far so good. Getting here was straightforward and IoTDB is easy to work with, it's a really cool piece of software!

## Where problems arise

The approach outlined above is straightforward and works well with StreamPipes.
However, there are two aspects we haven't considered yet that make things more challenging and are currently unclear how to support:
* Dimension properties
* Complex data types (mainly lists and nested data)

The latter is not directly relevant, but if anyone has a viable suggestion, I'd really appreciate it.
So let's have a deeper look on dimension properties.
In StreamPipes dimension properties refer to event properties (event fields) that represent dimensions.
In the example given above, `sensorId` would typically be modeled as a dimension property.
As such `sensorId` has discrete value space containing values like `flowrate01` and `flowrate02`.
Dimension properties allow users, e.g., to group data along the provided dimensions in the data explorer. See an example configuration from the StreamPipes UI in the screenshot below.
![image](https://github.com/apache/streampipes/assets/50115603/073b0011-ae6e-4649-978f-96be4d81fc78)

In general there can be more than one dimension property such as `locationId` and `sensorId` and users can potentially change an existing adapter and remove or add a dimension.

If our `flowrate` data stream is persisted as above, there is no way to group data based on `sensorId` best to my knowledge,
e.g., to calculate the average temperature per `sensorId`. Or is there anything I'm missing?

One possible solution to this, is to split the data of our data stream into multiple devices in IoTDB and work with tags.
The idea is now to have a device per value of a dimension property, in this case `flowrate.flowrate01` and `flowrate.flowrate02`.
This allows us to tag the corresponding time series, e.g., `flowrate.flowrate01.temperature` with a corresponding tag: `sensorId=flowrate01`.
This brings us the desired capability of grouping values along dimensions:
```sql
select avg(temperature) from root.streampipes.flowrate.** GROUP BY TAGS(sensorId)
```
![image](https://github.com/apache/streampipes/assets/50115603/1b3da5cd-ccd1-4538-8608-8e4e14fea48b)
Please excuse the different tag name and values in this screenshot, but I think you get the point.

As an alternative, we could also imagine to not use tags and use the `GROUP BY LEVEL` statement.
The good side here is, that this allows to calculate aggregations based on dimension values, it also has some downsides.
First of all, it would break with the current relationship of a data lake measure having one measurement in the time series storage.
Modeling as described here would result potentially in a huge amount of time series since there must be one for each combination of dimension property values.
In addition, it would make our queries more complex since we are not able to directly query all data for a data lake measure or to count all records for one data lake measure without further computations.

What are your thoughts about this considerations?
Do I use the concepts in the right way?
Are there any alternatives we could achieve the same results?
What if an adapter is modified? E.g., a dimension field is removed?

## What should be possible in the end

Beyond the scenario above we require the following functionalities to be performed via queries from IoTDB which may rise further considerations/issues and should be considered ideally in finding a solution:

- simply list all values per data lake measure (`Select *`)
- filter data based on values of dimension properties
- group data based on a dimension property (e.g., show a line series per dimension property in the data explorer)

**Comment by chrisdutz** [comment]
I was going to propose that for multiple devices there should be stored with different deviceIds in IoTDB. Also do I think that one device will most probably send one measurement using a fixed measure, so this ideally should be implemented as a tag assigned to a measure. 

While generally IoTDB can handle an unbounded amount of timeseries and an unbounded amount of measuements for each of these, my gut-feeling does tell me that option 1 would be the better way to approach this.

**Reply by chrisdutz** [reply]
However I do hope one of my colleagues will jump on this thread with more in detail help.

**Comment by ottlukas** [comment]
First of all thank you for getting that started.

One concept you did not mention are the "template" [Device Template](https://iotdb.apache.org/UserGuide/latest/SQL-Manual/SQL-Manual.html#device-template)  "IoTDB supports the device template function, enabling different entities of the same type to share metadata, reduce the memory usage of metadata, and simplify the management of numerous entities and measurements.".
As far as I understand this would be the "DataLakeMeasure".

+1 for using tags. this is IMHO the right path.
for your simply list, filter data and group by: [Select Clause](https://iotdb.apache.org/UserGuide/latest/SQL-Manual/SQL-Manual.html#select-clause) -> Aggregation query by one single / multiple tag

**Reply by bossenti** [reply]
Thanks for bringing device templates to the table @ottlukas!

If I understand the concept correctly, this would only help us if we have multiple data streams (or then data lake measures) in StreamPipes that share the same metadata, i.e. event properties. Right?

**Comment by JackieTien97** [comment]
Actually, we're developping on relational model, but it may need several months to be finished.

For current tree model of IoTDB, I suggest that you put all your tag dimensions' value into the path, like `root.streampipes.flowrate. flowrate02`. If you have more than one dimension, you can also keep adding that into path, like `root.streampipes.flowrate.West.flowrate02`.

Then, if you want to group the data for all dimensions, you can use `align by device` clause (https://iotdb.apache.org/UserGuide/latest/User-Manual/Query-Data.html#align-by-clause-1), like `select avg(temperature) from root.streampipes.flowrate.**`.

The only question for this `align by device` way is that it can only group data by all dimensions, if you only need one or some dimensions, you may need to do the aggregation in your client.

**Reply by bossenti** [reply]
Thank you @JackieTien97!
That's the way I'm thinking right now.
The only problem here is that the dimensions can change due to user adjustments. This results in the creation of new time series/path.
Do you have any idea how to overcome this?

**Reply by bossenti** [reply]
Is it already possible to test the current state of the relational model?

**Reply by JackieTien97** [reply]
relational model is at a very early stage, may be formally released at the end of this year.

**Reply by JackieTien97** [reply]
I wonder how you resolve this problem in InfluxDB, because once you change the tag value in InfluxDB, it will also generate a new series.

And actually, this can be easily resolved in our relational model, we have four types of columns: Time Column, ID Column(tag set in InfluxDB), Attribute Column and MeasurementColumn(field set in InfluxDB). And these changing dimensions can be saved in Attribute Column which can updated and belong to the previous device(won't create a new device).

**Reply by bossenti** [reply]
> I wonder how you resolve this problem in InfluxDB, because once you change the tag value in InfluxDB, it will also generate a new series.

Good point 😅 Might be the case that we have the same issue with InfluxDB?! @dominikriemer @tenthe

**Reply by bossenti** [reply]
> And actually, this can be easily resolved in our relational model, we have four types of columns: Time Column, ID Column(tag set in InfluxDB), Attribute Column and MeasurementColumn(field set in InfluxDB). And these changing dimensions can be saved in Attribute Column which can updated and belong to the previous device(won't create a new device).

Sounds promising 👍🏼

**Reply by tenthe** [reply]
There may be cases that don't work properly, but so far I haven't had any problems with it. In general, when writing data to the data lake, the user can decide whether the event schema should be overwritten or extended.

<img width="752" alt="grafik" src="https://github.com/apache/streampipes/assets/5279561/fba27ec2-256a-4066-a463-aebae6b32cfb">

The old data should always remain in influx, and the new data is written according to the new schema. 
The difference with influx is that the measurement does not change. For the IoTDB, we would have to adjust the path of the measurement.
How can we then access the data already stored? Do we need to migrate them or do we need to follow the legacy paths?

For influx, I believe the data is still available. But it is also possible that I am wrong here, are there other experiences with this?

**Reply by JackieTien97** [reply]
For IoTDB, old data still remain too. For example, if we decicde use level2 to represent the version of the device, like `root.db.version1.d1` and we can write data of that device into `root.db.version1.d1(time, s1, s2, s3)`. 

If version of that device upgrades to version2, we will generate another device(same as influx, it will generate another new seriesKey) called `root.db.version2.d1`, then we can write data into this new device `root.db.version2.d1(time, s1, s2, s3)`.
Old data of version1 are still in `root.db.version1.d1`.

You can use `select * from root.db.version1.d1` to query old data and use `select * from root.db.version2.d1` to query new data. Schema can also be different between `root.db.version1.d1` and `root.db.version2.d1`, like we have  `root.db.version1.d1(time, s1, s2, s3)`, but have `root.db.version1.d2(time, s4, s5, s6)`

**Reply by qiaojialin** [reply]
Hi, IoTDB's tree schema is Complete equivalence to InfluxDB's tag schema. Here is an example

<img width="1722" alt="image" src="https://github.com/apache/streampipes/assets/7240743/157ff8e6-33b8-4408-b51d-75fac84fc7e6">

Here are some concept correspondence:

Database (InfluxDB) = Database (IoTDB)
Measurment (InfluxDB) = Internal node under database (IoTDB)
Tags (InfluxDB) = Other internal nodes (IoTDB)
Fields (InfluxDB) = Leaf nodes (IoTDB)
Measurement + Tags (series in InfluxDB) = All internal nodes under database (device in IoTDB)

Adding new schema (Tag3 in the example) will create new series in InfluxDB, also IoTDB will create new devices (root.m1.A2.B2.C1, root.m1.A2.B3.C2). 

Adding new timeseries also do not affect old timeseries, we can still access data by its path in IoTDB.

**Reply by SteveYurongSu** [reply]
@bossenti @tenthe As previously mentioned, IoTDB will develop a relational model later on. I will also provide a development roadmap for the IoTDB relational model here. After the relational model is developed, the tree model and the relational model will coexist within IoTDB.

Relation Model V1 (expected to be released around July this year):
- Establish the overall framework of the table model engine.
- Writing: Support for JAVA SDK and SQL writing.
- Query: Single table raw data query, support for time and value filtering, common functions, and operators.
- Metadata: Database management, addition, deletion, and query of tables.

Relation Model V2 (expected to be released around September this year):
- Writing: Support for C++ and Python SDK writing, support for Java Session SDK redirection, support for compaction, TTL, deletion, and write performance that is not inferior to the original tree interface.
- Query: Support for multi-table inner join based on the time column, aggregate queries (group by clause support for any combination of time, identifier, and attribute columns), support for more scalar and aggregate functions.
- Metadata: Table modification, index creation, permissions.

I think we can start by adapting the tree model now. However, considering factors such as understandability, we can switch to adapting the relational model after the relational model is developed.

**Reply by bossenti** [reply]
@qiaojialin @JackieTien97 thank you very much for your detailed explanation! This underlings my current understanding of IoTDB's tree model.

@SteveYurongSu thank you, this looks awesome!

With the schedule in mind, I would suggest to wait for V1 of the relational model. It promises all the features and functionalities we currently have and I would like to save the implementation effort if we switch to the relational model either way. This way we can also be one of the early adopters of the new relational model and provide you with feedback.

**Reply by JackieTien97** [reply]
sure, look forward to your feedback
